src/cir.rs

```diff
@@ -0,0 +1,439 @@
+                return false;
+            } else {
+                if is_index(&and.right) {
+                    return execute_helper(cir_instructions, and.right.as_index().unwrap(), ctx, m);
+                } else {
+                    return and.right.as_predicate().unwrap().execute(ctx, m);
+                }
+            }
+        }
+        CirInstruction::OrIns(or) => {
+            let left_val = if is_index(&or.left) {
+                execute_helper(cir_instructions, or.left.as_index().unwrap(), ctx, m)
+            } else {
+                or.left.as_predicate().unwrap().execute(ctx, m)
+            };
+
+            if left_val {
+                // short circuit
+                return true;
+            } else {
+                if is_index(&or.right) {
+                    return execute_helper(cir_instructions, or.right.as_index().unwrap(), ctx, m);
+                } else {
+                    return or.right.as_predicate().unwrap().execute(ctx, m);
+                }
+            }
+        }
+        CirInstruction::NotIns(not) => {
+            let right_val = if is_index(&not.right) {
+                execute_helper(cir_instructions, not.right.as_index().unwrap(), ctx, m)
+            } else {
+                not.right.as_predicate().unwrap().execute(ctx, m)
+            };
+            return !right_val;
+        }
+    }
+}
+
+impl Execute for CirProgram {
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
+        execute_helper(&self.instructions, self.instructions.len() - 1, ctx, m)
+    }
+}
+
+fn add_to_counter_helper(
+    cir_instructions: &[CirInstruction],
+    index: usize,
+    map: &mut HashMap<String, usize>,
+) {
+    match &cir_instructions[index] {
+        CirInstruction::AndIns(and) => {
+            if !is_index(&and.left) {
+                *map.entry(and.left.as_predicate().unwrap().lhs.var_name.clone())
+                    .or_default() += 1;
+            }
+            if !is_index(&and.right) {
+                *map.entry(and.right.as_predicate().unwrap().lhs.var_name.clone())
+                    .or_default() += 1;
+            }
+        }
+        CirInstruction::OrIns(or) => {
+            if !is_index(&or.left) {
+                *map.entry(or.left.as_predicate().unwrap().lhs.var_name.clone())
+                    .or_default() += 1;
+            }
+            if !is_index(&or.right) {
+                *map.entry(or.right.as_predicate().unwrap().lhs.var_name.clone())
+                    .or_default() += 1;
+            }
+        }
+        CirInstruction::NotIns(not) => {
+            if !is_index(&not.right) {
+                *map.entry(not.right.as_predicate().unwrap().lhs.var_name.clone())
+                    .or_default() += 1;
+            }
+        }
+    }
+}
+
+fn remove_from_counter_helper(
+    cir_instructions: &[CirInstruction],
+    index: usize,
+    map: &mut HashMap<String, usize>,
+) {
+    match &cir_instructions[index] {
```

Comment:
Same question here, we should iterator the `cir_instructions` to reduce counter for every field